src/org/opensolaris/opengrok/configuration/messages/MessageListener.java

```diff
@@ -0,0 +1,435 @@
+     */
+    public void start(SocketAddress endpoint) throws IOException {
+        if (isListening()) {
+            throw new IllegalStateException("Cannot start when already listening for messages");
+        }
+        bindToAddress(endpoint);
+        new Thread(() -> {
+            while (!configServerSocket.isClosed()) {
+                acceptConnection();
+            }
+        }, "configurationListener").start();
+    }
+
+    private boolean isListening() {
+        return configServerSocket != null && !configServerSocket.isClosed();
+    }
+
+    private void bindToAddress(final SocketAddress address) throws IOException {
+        configServerSocket = new ServerSocket();
+        configServerSocket.bind(address);
+    }
+
+    private void acceptConnection() {
+        try (Socket s = configServerSocket.accept();
+             MessageReader in = new MessageReader(s.getInputStream());
+             OutputStream output = s.getOutputStream()) {
+
+            LOGGER.log(Level.FINE, "OpenGrok: Got request from {0}", s.getInetAddress().getHostAddress());
+
+            Message m = in.readMessage();
+            handleMessage(m, output);
+
+        } catch (SocketException e) {
+            // if closed then it is okay (socket was closed while blocked in accept())
+            if (!configServerSocket.isClosed()) {
+                LOGGER.log(Level.SEVERE, "Socket error", e);
+            }
+        } catch (IOException e) {
+            LOGGER.log(Level.SEVERE, "Error reading config file: ", e);
+        } catch (RuntimeException e) {
+            LOGGER.log(Level.SEVERE, "Error parsing config file: ", e);
+        } catch (ValidationException e) {
+            LOGGER.log(Level.WARNING, "Received invalid message", e);
+        }
+    }
+
+    /**
+     * Handle incoming message.
+     *
+     * @param m message
+     * @param output output stream for errors or success
+     */
+    private void handleMessage(final Message m, final OutputStream output) throws IOException, ValidationException {
+        m.validate();
+
+        if (!canAcceptMessage(m)) {
+            LOGGER.log(Level.WARNING, "Message dropped: {0} - too many messages in the system", m.getTags());
+            output.write(Message.DeliveryStatus.OVER_LIMIT.getStatusCode());
+            return;
+        }
+
+        Response response;
+        try {
+            response = processMessage(m);
+        } catch (Exception ex) {
+            LOGGER.log(Level.WARNING, String.format("Message dropped: %s - message error", m), ex);
+            output.write(Message.DeliveryStatus.ERROR.getStatusCode());
+            output.write(ex.getMessage().getBytes());
+            return;
+        }
+
+        LOGGER.log(Level.FINER, "Message received: {0}", m.getTags());
+        LOGGER.log(Level.FINER, "Messages in the system: {0}", getMessagesInTheSystem());
+
+        output.write(Message.DeliveryStatus.OK.getStatusCode());
+        output.write(gson.toJson(response).getBytes());
```

Comment:
can you abstract the encoding/decoding ? `message.getEncoded()` ?

Reply:
What do you mean exactly? Messages encoding and decoding is in different part of the code. Either in `Message` class or in `MessageReader/MessageWriter`. This part just sends back the generic `Response`.

Reply:
I mean - could the `gson` part be constrained just to `Message` class ? i.e. is it necessary to call gson directly in this place ?

Reply:
Yes, because I am marshalling `Response` and not `Message`. To clear things up, you just want me to add method `getEncoded()` to `Response` class so there would not have to be `gson` field in the `MessageListener`, right?

Reply:
I.e. to make it easily replaceable in the future with different encoding mechanism (once JSON ceases being hip).